### PR TITLE
Changes to sample data editing

### PR DESCRIPTION
**Pull request by zsloan**
This PR will make a couple changes to sample data editing:

- Values will be editable for samples that don't currently have values. Currently you can only edit sample values for samples that already have values; this is because it just pulls the sample data from the DB and wasn't accounting for the sample list
- It will be possible to submit sample data changes via an HTML table, instead of just as a CSV import

The accompanying GN2 PR is here - https://github.com/genenetwork/genenetwork2/pull/781

### CODE REVIEW

File: gn3/db/datasets.py [hunk 1]
```diff
@@ -1,8 +1,57 @@
...
 
+def retrieve_sample_list(group: str):
+    """
+    Get the sample list for a group (a category that datasets belong to)
+
+    Currently it is fetched from the .geno files, since that's the only place
+    the "official" sample list is stored
+    """
+
+    genofile_path = os.environ.get("GENENETWORK_FILES") + "/genotype/" + group + ".geno"
```
Comment (BonfaceKilz) [inline gn3/db/datasets.py:15]: *nitpick:* genofile_path is a constant, and as such, upper-case it.

File: gn3/db/datasets.py [hunk 2]
```diff
@@ -1,8 +1,57 @@
...
+                continue
+            break
+
+        headers = line.split("\t")
+
+        if headers[3] == "Mb":
+            samplelist = headers[4:]
+        else:
+            samplelist = headers[3:]
+        return samplelist
```
Comment (BonfaceKilz) [inline gn3/db/datasets.py:33]: *issue:* When opening a file, use a context-manager, in this case, `with open ...`.  This way, you don't have to worry about closing any hanging file connections should something unforeseen happens.

File: gn3/db/datasets.py [hunk 3]
```diff
@@ -1,8 +1,57 @@
...
+def retrieve_group_name(
+        group_id: int, connection: any):
+    """
+    Given the group id (InbredSet.Id in the database), retrieve its name
+    """
+    query = (
+        "SELECT Name "
+        "FROM InbredSet "
+        "WHERE "
+        "InbredSet.Id = %(group_id)s")
```
Comment (BonfaceKilz) [inline gn3/db/datasets.py:44]: *question:* is this escaped?  I thought just using simple "%s" like so:

``` "SELECT * FROM .... WHERE ... = %s", value```

should escape things.

File: gn3/db/datasets.py [hunk 4]
```diff
@@ -1,8 +1,57 @@
...
+    with connection.cursor() as cursor:
+        cursor.execute(
+            query,
+            {
+                "group_id": group_id
+            })
+        res = cursor.fetchone()
+        if res:
+            return res[0]
+        return None
```
Comment (BonfaceKilz) [inline gn3/db/datasets.py:54]: *nitpick:* @arunisaac write some monadic constructs to deal with None values when doing SQL queries.  You could look at those.  If you get stuck, we could chat over on Matrix and I could show you how to use them.

**Review by BonfaceKilz — APPROVED** [review]
@zsloan thanks for pinging me to review this.  I've left some minor comments that you could look at.  Also, generally, you could use tests to try and poke holes at weird edge cases.

**Review by BonfaceKilz — APPROVED** [review]
@zsloan thanks for pinging me to review this.  I've left some minor comments that you could look at.  Also, generally, you could use tests to try and poke holes at weird edge cases.
